source/Halibut/Util/RetryPolicy.cs

```diff
@@ -0,0 +1,63 @@
+using System;
+using System.Diagnostics;
+
+namespace Halibut.Util
+{
+    class RetryPolicy
+    {
+        readonly Stopwatch stopwatch = new Stopwatch();
+
+        internal RetryPolicy(double backoffMultiplier, TimeSpan minimumDelay, TimeSpan maximumDelay)
+        {
+            if (backoffMultiplier <= 0) throw new ArgumentOutOfRangeException(nameof(backoffMultiplier), "Must be greater than zero");
+            if (backoffMultiplier > 1) throw new ArgumentOutOfRangeException(nameof(backoffMultiplier), "Must be less than or equal to one");
+
+            BackoffMultiplier = backoffMultiplier;
+            MinimumDelay = minimumDelay;
+            MaximumDelay = maximumDelay;
+        }
+
+        internal TimeSpan MinimumDelay { get; }
+        internal TimeSpan MaximumDelay { get; }
+        internal double BackoffMultiplier { get; }
+
+        public static RetryPolicy Create()
+        {
+            var backoffMultiplier = new Random().NextDouble();
+            var minimumDelay = TimeSpan.FromSeconds(5);
+            var maximumDelay = TimeSpan.FromMinutes(2);
+            return new RetryPolicy(backoffMultiplier, minimumDelay, maximumDelay);
+        }
+
+        public void Try()
+        {
+            stopwatch.Start();
+        }
+
+        public void Success()
+        {
+            stopwatch.Reset();
+        }
+
+        public TimeSpan GetSleepPeriod()
+        {
+            var elapsed = stopwatch.Elapsed;
+            var delay = CalculateDelay(elapsed);
+            return delay;
+        }
+
+        internal TimeSpan CalculateDelay(TimeSpan elapsed)
+        {
+            var multiple = 1d + BackoffMultiplier;
+            var delay = TimeSpan.FromMilliseconds(elapsed.TotalMilliseconds * multiple);
+            delay = delay > MinimumDelay
```

Comment:
This is a good way of making sure your value is within a range 👍 (I would have most likely done some nested-ternary nastiness to jam it all on one line, but I really like this approach) :pipe-thinking-wizard:

Reply:
Thank you :)

For what it's worth, under the covers the compiler will optimise it to look pretty much what the nested ternary would have done anyway. I'm happy to optimise for humans to read and let the compiler optimise it for machines to read :)